packages/matplotlib/patches/0004-WebAgg-backend.patch

```diff
@@ -0,0 +1,721 @@
++            alert(f"Cannot download plot, unable to determine mimetype for '{format}'")
++            return
++
++        element = document.createElement('a')
++        data = BytesIO()
++        self.figure.savefig(data, format=format)
++
++        element.setAttribute(
++            "href",
++            "data:{};base64,{}".format(
++                mimetype, base64.b64encode(data.getvalue()).decode("ascii")
++            ),
++        )
++        element.setAttribute("download", f"plot{figure_id}.{format}")
++        element.style.display = "none"
++        document.body.appendChild(element)
++        element.click()
++        document.body.removeChild(element)
++
+     def send_event(self, event_type, **kwargs):
+         if self.manager:
+             self.manager._send_event(event_type, **kwargs)
+@@ -498,6 +473,10 @@ class FigureManagerWebAgg(backend_bases.FigureManagerBase):
+         output.write("mpl.default_extension = {};".format(
+             json.dumps(FigureCanvasWebAggCore.get_default_filetype())))
+
++        output.write("mpl.toolbar_image_callback = null;\n")
++        output.write("mpl.set_toolbar_image_callback = function(c) {mpl.toolbar_image_callback=c;}\n")
++        output.write("mpl.set_toolbar_image_callback;\n")
++
+         if stream is None:
+             return output.getvalue()
+
+diff --git a/lib/matplotlib/backends/web_backend/js/mpl.js b/lib/matplotlib/backends/web_backend/js/mpl.js
+index 140f590385..1776b1ff4e 100644
+--- a/lib/matplotlib/backends/web_backend/js/mpl.js
++++ b/lib/matplotlib/backends/web_backend/js/mpl.js
+@@ -2,19 +2,58 @@
+ /* global mpl */
+ window.mpl = {};
+
+-mpl.get_websocket_type = function () {
+-    if (typeof WebSocket !== 'undefined') {
+-        return WebSocket;
+-    } else if (typeof MozWebSocket !== 'undefined') {
+-        return MozWebSocket;
+-    } else {
+-        alert(
+-            'Your browser does not have WebSocket support. ' +
+-                'Please try Chrome, Safari or Firefox ≥ 6. ' +
+-                'Firefox 4 and 5 are also supported but you ' +
+-                'have to enable WebSockets in about:config.'
+-        );
++class MockJsWebSocket {
++    binaryType = "blob";
++
++    constructor(fig_id) {
++        this.fig_id = fig_id;
++        this.readyState = 0;
++        this.python_web_socket = null;
++        this.python_onmessage_callback = null;
++    }
++
++    get onopen() {
++        return this._onopen;
++    }
++
++    set onopen(listener) {
++        this._onopen = listener;
++        this.readyState = 1;
+     }
++
++    open(python_onmessage_callback) {
++        this.python_onmessage_callback = python_onmessage_callback;
++        if (this.onopen != null) {
++            this.onopen();
++        }
++    }
++
++    receive_binary(content, binary=true) {
++        var buffer = content.getBuffer();
++        content.destroy();
++        try {
++            blob = new Blob([buffer.data]);
++            this.onmessage({ data: blob });
```

Comment:
It doesn't look like we declared blob? Also if we call it data we can shorten a bit.
```suggestion
+            const data = new Blob([buffer.data]);
+            this.onmessage({ data });
```

Reply:
Done